gobblin-core-base/src/main/java/gobblin/converter/AnyToJsonRecordWithMetadataConverter.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package gobblin.converter;
+
+import java.util.Collections;
+import java.util.Map;
+
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericData;
+import org.apache.avro.generic.GenericRecord;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonElement;
+
+import gobblin.configuration.WorkUnitState;
+import gobblin.type.RecordWithMetadata;
+
+
+/**
+ * A converter that converts any type to a {@link RecordWithMetadata}.
+ * Ensures that the contained Record is a JsonElement. If the recordType can be retrieved from the
+ * object or inferred from the input schema, it will be placed in metadata.
+ * Caveat: Currently only supports Avro and Json input
+ */
+public class AnyToJsonRecordWithMetadataConverter extends Converter<Object, String, Object, RecordWithMetadata<JsonElement>> {
+  private static final Gson GSON = new GsonBuilder().disableHtmlEscaping().create();
+  private String recordTypeName;
+
+  @Override
+  public String convertSchema(Object inputSchema, WorkUnitState workUnit)
+      throws SchemaConversionException {
+    // Store the recordName if we know what it is
+    if (recordTypeName == null && inputSchema != null && inputSchema instanceof Schema) {
+      recordTypeName = ((Schema) inputSchema).getFullName();
+    }
+    return "";
+  }
+
+  @Override
+  public Iterable<RecordWithMetadata<JsonElement>> convertRecord(String outputSchema, Object inputRecord,
+      WorkUnitState workUnit)
+      throws DataConversionException {
+    if (inputRecord instanceof RecordWithMetadata && ((RecordWithMetadata) inputRecord)
```

Comment:
Realized something that we should probably make a pattern for these input-type-adjusting converters. 

Since converter chains are generally static, you can lock down the input type and therefore the cast pattern after the first record passes through. This will allow us to skip instanceof checks after the first record passes through. 

Maybe with a config flag to revert to the unoptimized behavior if the pipeline author has set up a pipeline that can generate different types of records. 

We don't have to address this in this PR, but might be worth picking up as a follow-up PR. Thoughts?